### ISSUE TITLE
Issues with the C++ Tuning API

### ISSUE THREAD

**Issue by u235axe** [CNugteren/CLBlast#315]
Hi!

I've tried the tuning feature and run into problems (Windows 10, 64bit build, unchanged CMAKE settings file, VS2017, AMD GPU, Intel Integrated GPU).

0) The tuning sample code works as expected with the provided float version.
1) If I change it to 'double', TuneCopy fails with code  -51 (kInvalidArgSize).

I tried to tune gemv also, similar things happen:
2) For floats the tuning finishes, meaningful parameters are printed from the unordered map, but OverrideParameters cannot set them, it fails with -2047 (kMissingOverrideParameter)
3) If I try to tune gemv for doubles, it fails with code -51 like in the copy case.

Any ideas what could have gone wrong?

**Comment by CNugteren** [comment]
Sorry, I'm not really sure what you are doing. First of all, what is the "tuning sample code"? You mean the actual tuners? Secondly, when you say "change it to double", do you mean you are editing the code? Probably you can better set the command-line arguments (see the help message when the tuners start).

About the `OverrideParameters` error, make sure you match the arguments with what you'll find in the database, e.g. see [here](https://github.com/CNugteren/CLBlast/blob/master/src/database/kernels/xgemv/xgemv_32.hpp#L14) for the first GEMV kernel (note that there are 3).

**Comment by CNugteren** [comment]
Ah, OK, now everything is clear. It is a relatively new feature, so there could be issues.

I've tried to reproduce:
* I just compiled & ran `clblast_sample_tuning_api` and everything seems fine.
* Then I changed `float` into `double` in [this line](https://github.com/CNugteren/CLBlast/blob/master/samples/tuning_api.cpp#L63) and `kSingle` into `kDouble` in [this line](https://github.com/CNugteren/CLBlast/blob/master/samples/tuning_api.cpp#L72), compiled again, and it still runs fine. If this doesn't help you, perhaps you can run the regular tuner executable first so you'll get a bit more info for debugging this issue.
* I went back to single-precision and then change to `TuneXgemv` in [this line](https://github.com/CNugteren/CLBlast/blob/master/samples/tuning_api.cpp#L63) and `Gemv` in [this line](https://github.com/CNugteren/CLBlast/blob/master/samples/tuning_api.cpp#L72), compiled again and now I got:
```
Starting the tuning...
Completed TuneXgemv with status 0 (0 == OK), found parameters:
>  WPT3 = 16
>  WGS3 = 32
>  VW3 = 8
>  WPT2 = 1
>  VW2 = 1
>  WGS1 = 32
>  WPT1 = 1
>  WGS2 = 16
CLBlast: Run-time error: -2041
Completed OverrideParameters with status -2041 (0 == OK)
```
So tuning succeeds, but the `OverrideParameters` doesn't like it. I believe it needs to split up in 3 parts, it is a bit of a special case actually... see [here](https://github.com/CNugteren/CLBlast/blob/master/src/database/kernels/xgemv/xgemv_32.hpp#L14), [here](https://github.com/CNugteren/CLBlast/blob/master/src/database/kernels/xgemv_fast/xgemv_fast_32.hpp#L14), and [here](https://github.com/CNugteren/CLBlast/blob/master/src/database/kernels/xgemv_fast_rot/xgemv_fast_rot_32.hpp#L14). This should be fixed or at least documented better... my apologies.

**Comment by CNugteren** [comment]
I have decided to just document this issue rather than to change the behaviour, mostly because this is for an expert-user setting anyway. You can find the documentation here:
https://github.com/CNugteren/CLBlast/blob/master/doc/tuning.md#inspecting-and-changing-tuning-parameters-at-run-time

Does this solve the issue for you?

**Comment by u235axe** [comment]
Yes, the regular tuner works fine for me also.

So if I get it right, OverrideParameters cannot ignore the extra parameter entries returned by the tuner API and I need to filter them manually?

As per the table in the docs, gemv needs UNROLL also, but I haven't seen that anywhere even in the result of the tuning, am I missing something?

**Comment by CNugteren** [comment]
I know it has been a while, but I did make a fix today in the `CLBlast-315-tuning-api-improvements` branch (soon to be merged into master). I've removed the UNROLL parameters from the documentation (thanks for pointing it out) and I've made the tuning API flexible as you requested, it disregards any extra parameter values.

So, modifying the sample for GEMV would give you [this code](https://gist.github.com/CNugteren/d7e3626325c6f8f3a07c9858221919d6). Hope that's what you were looking for and I hope it's not too late.
